src/main/scala/net/psforever/actors/session/support/SessionData.scala

```diff
@@ -1640,92 +1403,392 @@ class SessionData(
+    session.player.spectator = spectator
+  }
+
+  def handleKick(player: Player, time: Option[Long]): Unit = {
+    administrativeKick(player)
+    accountPersistence ! AccountPersistenceService.Kick(player.Name, time)
+  }
+
+  def handleSilenced(isSilenced: Boolean): Unit = {
+    player.silenced = isSilenced
+  }
+
+  /* supporting functions */
+
+  val sessionSupervisorStrategy: SupervisorStrategy = {
+    import net.psforever.objects.inventory.InventoryDisarrayException
+    OneForOneStrategy(maxNrOfRetries = -1, withinTimeRange = 1 minute) {
+      case nge: NoGUIDException =>
+        nge.getEntity match {
+          case p: Player =>
+            continent.GUID(p.VehicleSeated) match {
+              case Some(v: Vehicle) =>
+                attemptRecoveryFromNoGuidExceptionAsVehicle(v, nge)
+              case _ =>
+                attemptRecoveryFromNoGuidExceptionAsPlayer(p, nge)
+            }
+
+          case v: Vehicle =>
+            attemptRecoveryFromNoGuidExceptionAsVehicle(v, nge)
+
+          case e: Equipment =>
+            (
+              player.Holsters().zipWithIndex.flatMap { case (o, i) =>
+                o.Equipment.flatMap { item => Some((player, InventoryItem(item, i))) }
+              }.toList ++
+                player.Inventory.Items.map { o => (player, o) } ++
+                player.FreeHand.Equipment.flatMap { item => Some((player, InventoryItem(item, Player.FreeHandSlot))) }.toList ++
+                (validObject(player.VehicleSeated) match {
+                  case Some(v: Vehicle) => v.Trunk.Items.map { o => (v, o) }
+                  case _ => Nil
+                })
+              )
+              .find { case (_, InventoryItem(o, _)) => o eq e } match {
+              case Some((p: Player, InventoryItem(obj, index))) if !obj.HasGUID =>
+                p.Slot(index).Equipment = None
+                attemptRecoveryFromNoGuidExceptionAsPlayer(p, nge)
+              case Some((p: Player, _)) =>
+                attemptRecoveryFromNoGuidExceptionAsPlayer(p, nge)
+
+              case Some((v: Vehicle, InventoryItem(obj, index))) if !obj.HasGUID && v.PassengerInSeat(player).contains(0) =>
+                v.Slot(index).Equipment = None
+                attemptRecoveryFromNoGuidExceptionAsPlayer(player, nge)
+                SupervisorStrategy.resume
+              case Some((v: Vehicle, InventoryItem(obj, index))) if !obj.HasGUID =>
+                v.Slot(index).Equipment = None
+                SupervisorStrategy.resume
+              case Some((v: Vehicle, _)) if v.PassengerInSeat(player).contains(0) =>
+                attemptRecoveryFromNoGuidExceptionAsPlayer(player, nge)
+                SupervisorStrategy.resume
+              case Some((_: Vehicle, _)) =>
+                SupervisorStrategy.resume
+
+              case _ =>
+                //did not discover or resolve the situation
+                writeLogExceptionAndStop(nge)
+            }
```

Comment:
not 100% sure but it might be possible to combine those into a `collect` and that might make it cleaner

Reply:
I am uncertain if that will improve much.  None of the cases would change; and, I only want one result anyway.

Reply:
well it reads weird to have a `chain(...).chain(...).chain(...) match {...}` instead of `chain.chain.chain`